core/src/test/java/io/aiven/kafka/tieredstorage/transform/RateLimitedInputStreamTest.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2024 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.transform;
+
+import java.io.ByteArrayInputStream;
+import java.io.InputStream;
+import java.time.Duration;
+import java.util.Arrays;
+
+import io.github.bucket4j.Bucket;
+import org.junit.jupiter.api.Test;
+
+import static org.awaitility.Awaitility.await;
+
+class RateLimitedInputStreamTest {
+    @Test
+    void testDoesNotBlockRead() {
+        // Given a bucket with min size of default buffer
+        final Bucket bucket = RateLimitedInputStream.rateLimitBucket(1);
+        // When a stream of size less than capacity is read
+        final byte[] bytes = new byte[RateLimitedInputStream.DEFAULT_BUFFER_SIZE - 1];
+        final ByteArrayInputStream source = new ByteArrayInputStream(bytes);
+        final InputStream test = new RateLimitedInputStream(source, bucket);
+        // Then read should happen without blocking, i.e. less than 1 sec
+        await().atMost(Duration.ofSeconds(1))
+            .until(() -> {
+                test.readAllBytes();
+                return true;
+            });
```

Comment:
Technically, this and similar tests below have some potential for flakiness. We could probably circumvent this by passing a mock of `BlockingStrategy` to `consume`. I'm fine if we leave it as is, but it would be good to make an issue about this.

Reply:
Sure, adding #551